### PR TITLE
Record HTTP requests/response communication in integration tests

### PR DESCRIPTION
**Pull request by jirikuncar**
# TODO
- [x] `tests/integration/api/test_api.py::TestDatadog::test_metrics SKIPPED     [ 13%]`
- [x] `tests/integration/dogshell/test_dogshell.py::TestDogshell::test_monitors` update the error message
- [ ] remove `click` from dependencies (use `capsys` and `except SystemExit`): **Q: Is it really necessary to remove it?**

### What does this PR do?

Records/replays HTTP communication of integration tests.

### Description of the Change

- ⏺ `--vcr-record=all` 
- ⏯ `--vcr-record=none`

### Possible Drawbacks

The size of cassettes might grow.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: tests/integration/api/test_api.py [hunk 1]
```diff
@@ -206,14 +207,17 @@ def test_timeboard(self):
...
-    def test_search(self):
+    def test_search(self, dog):
         results = dog.Infrastructure.search(q="")
         assert len(results["results"]["hosts"]) > 0
         assert len(results["results"]["metrics"]) > 0
 
-    def test_metrics(self):
-        now = datetime.datetime.now()
-        now_ts = int(time.mktime(now.timetuple()))
+    @pytest.mark.skip
```
Comment (jirikuncar) [inline tests/integration/api/test_api.py]: @DataDog/integrations-tools-and-libraries I am having issues with recording and replaying this test case 👀

File: .azure-pipelines/integration_tests.yml [hunk 2]
```diff
@@ -16,13 +16,13 @@ jobs:
       versionSpec: "3.7"
   - script: pip install tox
     displayName: Install tox
-  - script: tox -e integration
+  - script: tox -e integration -- --vcr-record=all
```
Comment (jirikuncar) [inline .azure-pipelines/integration_tests.yml]: Integration tests are using real API endpoints

File: datadog/api/api_client.py [hunk 3]
```diff
@@ -145,7 +146,7 @@ def submit(cls, method, path, api_version=None, body=None, attach_host_name=Fals
 
             # Process the body, if necessary
             if isinstance(body, dict):
-                body = json.dumps(body)
+                body = json.dumps(body, sort_keys=cls._sort_keys)
```
Comment (jirikuncar) [inline datadog/api/api_client.py]: The body has to be stable when compressed for `body` matcher.

File: datadog/dogshell/common.py [hunk 4]
```diff
@@ -16,6 +16,7 @@ def print_err(msg):
         print(msg + '\n', file=sys.stderr)
     else:
         sys.stderr.write(msg + '\n')
+    sys.stderr.flush()
```
Comment (jirikuncar) [inline datadog/dogshell/common.py]: make sure it's printed when on buffered io

File: tests/integration/api/test_aws_integration.py [hunk 5]
```diff
@@ -20,25 +19,14 @@
...
-        """
-        initialize(api_key=API_KEY, app_key=APP_KEY, api_host=API_HOST)
-
-        for acc in product(ACCOUNT_IDS, ROLE_NAMES):
-            cls.accounts_to_cleanup.append({'account_id': acc[0], 'role_name': acc[1]})
-
-    def teardown_method(self, method):
-        for account in self.accounts_to_cleanup:
-            dog.AwsIntegration.delete(account_id=account['account_id'], role_name=account.get('role_name'))
+    @pytest.fixture(autouse=True)  # TODO , scope="class"
```
Comment (jirikuncar) [inline tests/integration/api/test_aws_integration.py]:
```suggestion
    @pytest.fixture(autouse=True)
```

I think it should be executed after each test case so the `TODO` can be removed

File: tests/integration/api/utils.py [hunk 6]
```diff
@@ -1,34 +0,0 @@
-# Unless explicitly stated otherwise all files in this repository are licensed under the BSD-3-Clause License.
```
Comment (jirikuncar) [inline tests/integration/api/utils.py]: moved to `conftest.py`

File: tests/integration/dogshell/test_dogshell.py [hunk 7]
```diff
@@ -31,54 +32,156 @@ def get_temp_file():
...
+    filename = tmp_path / ".test.dog.ini"
+    with filename.open("w") as fp:
+        dogshell_config.write(fp)
+    return str(filename)
+
+
+@pytest.fixture
+def dogshell(capsys, config_file, dog):
+    """Helper function to call the dog shell command."""
+    import click
```
Comment (jirikuncar) [inline tests/integration/dogshell/test_dogshell.py]: If needed I can refactor it to capture `stdout` and `stderr` manually. It's just a pain to keep it PY2/3 compatible.

File: tests/integration/test_freezer.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# Unless explicitly stated otherwise all files in this repository are licensed under the BSD-3-Clause License.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2015-Present Datadog, Inc
+
+import datetime
+
+
+def test_freezer(freezer):
+    with freezer:
+        assert datetime.datetime.now() == freezer.time_to_freeze
```
Comment (jirikuncar) [inline tests/integration/test_freezer.py]: quickly discover issues with `freezegun`

File: .azure-pipelines/unit_tests.yml [hunk 9]
```diff
@@ -48,3 +48,11 @@ jobs:
     displayName: Install tox
   - script: tox
     displayName: Run unit tests
+  - script: tox -e integration -- --vcr-record=none
+    displayName: Run integration tests on cassettes
+    env:
+      DD_TEST_CLIENT_USER: frog@datadoghq.com
```
Comment (zippolyte) [inline .azure-pipelines/unit_tests.yml]: Let's store that in a variable in azure

File: tests/integration/api/test_api.py (hunk 1 above)
Comment (zippolyte) [reply tests/integration/api/test_api.py]: What kind of issues ?

File: .azure-pipelines/unit_tests.yml (hunk 9 above)
Comment (jirikuncar) [reply .azure-pipelines/unit_tests.yml]: the problem is that it must be a constant that was used for recording

Comment (zippolyte) [reply .azure-pipelines/unit_tests.yml]: Not sure why that's a problem

Comment (jirikuncar) [reply .azure-pipelines/unit_tests.yml]: If I don't keep it here, then I should store it together with cassettes similar to `seed` or `frozen` time.

Comment (jirikuncar) [reply .azure-pipelines/unit_tests.yml]: I am storing the `user_handle` per test case together with cassettes.

**Pull request merged**